### PR TITLE
Switched to use helix 0.6.x for Gobblin on Yarn

### PR DESCRIPTION
**Pull request by liyinan926**
Main changes:

1. Switched to Helix `0.6.6-SNAPSHOT`.
2. Added implementation of `GobblinHelixJobLauncher.executeCancellation`.
3. Other fixes and code refactorings.

Signed-off-by: Yinan Li <liyinan926@gmail.com>

### CODE REVIEW

**Comment by liyinan926** [comment]
@zliu41 can you review?

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 1]
```diff
@@ -487,8 +487,10 @@ private static void cleanupStagingDataPerTask(JobState jobState) {
     Map<String, ParallelRunner> parallelRunners = Maps.newHashMap();
     try {
       for (TaskState taskState : jobState.getTaskStates()) {
+        TaskState fatTaskState = new TaskState(taskState);
+        fatTaskState.addAllIfNotExist(jobState);
```
Comment (zliu41) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: This could be slow if the number of tasks is large. Can you just add the properties needed?

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java [hunk 2]
```diff
@@ -111,12 +114,35 @@ public GobblinHelixJobLauncher(Properties jobProps, HelixManager helixManager, P
...
+    } finally {
+      super.close();
+    }
+  }
+
+  @Override
   protected void runWorkUnits(List<WorkUnit> workUnits) throws Exception {
     try {
       submitJobToHelix(createJob(workUnits));
+      LOGGER.info(String.format("Submitted job %s to Helix", this.jobContext.getJobId()));
```
Comment (zliu41) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java]: Should this line and the next line be synchronized?

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java [hunk 3]
```diff
@@ -111,12 +114,35 @@ public GobblinHelixJobLauncher(Properties jobProps, HelixManager helixManager, P
   }
 
   @Override
+  public void close() throws IOException {
+    try {
+      if (this.jobSubmitted && !this.jobComplete) {
```
Comment (zliu41) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java]: You can call `executeCancellation` here

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java [hunk 4]
```diff
@@ -121,28 +122,34 @@ public TaskResult run() {
...
   @Override
   public void cancel() {
-    LOGGER.info(String.format("Cancelling WorkUnit %s of job %s", this.taskId, this.jobId));
+    LOGGER.info(String.format("Cancelling task %s of job %s", this.taskId, this.jobId));
     this.futureOptional.transform(new Function<Future<?>, Object>() {
       @Nullable
       @Override
       public Object apply(Future<?> input) {
-        return input.cancel(true);
+        return input.cancel(false);
```
Comment (zliu41) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: what's the reason of this change?

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java (hunk 1 above)
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: We are doing the same in `MRJobLauncher`. I will try to fix both places.

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java (hunk 4 above)
Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: To allow already-started tasks to finish gracefully.

Comment (zliu41) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: If an `InterruptedException` is caught in line 133, shouldn't the task be cancelled asap? Also what about hanging tasks?

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java (hunk 2 above)
Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java]: No, this method will only be called from a single thread.

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java (hunk 3 above)
Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java]: Fixed.

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java (hunk 4 above)
Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: It all depends on if the task responds to interruption or not. But you are right, if an InterruptedException is caught in line 133, then cancellation should be immediate. I will change it back.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java (hunk 1 above)
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: Actually seems there's a bunch of config properties needed so it might be a good idea to keep the logic here instead of iterating through every possible config property. I think a better approach would be to put `fatTaskState.addAllIfNotExist(jobState);` into the `ParallelRunner`.

Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: It seems it's pretty tricky to add this into `ParallelRunner`.  I think this should be fine because this is entirely in memory, so it should be much less significant than deleting files from HDFS.

**Comment by liyinan926** [comment]
@zliu41 Updated  d0774f6.

**Comment by zliu41** [comment]
LGTM
